finance_dl/gemini.py

```diff
@@ -0,0 +1,241 @@
+    trades = []
+    tprev = None
+    ts = int(time.time() - 3600*24*365*20) * 1000 # start 20 years ago
+    t = datetime.datetime.now()
+    payload_nonce =  int(time.mktime(t.timetuple())*1000)
+
+    counter = 0
+    while ts != tprev:
+        print(counter)
+        tprev = ts
+        payload =  {"request": "/v1/mytrades", 
+                    "nonce": str(payload_nonce), 
+                    "limit_trades": 500, 
+                    "timestamp" : ts}
+        encoded_payload = json.dumps(payload).encode()
+        b64 = base64.b64encode(encoded_payload)
+        signature = hmac.new(api_secret, b64, hashlib.sha384).hexdigest()
+        request_headers = get_request_headers(api_key, b64, signature)
+        response = requests.post(TRADES_URL, headers=request_headers)
+        response.raise_for_status()
+        my_trades = response.json()
+        if 'result' in my_trades:
+            raise Exception(my_trades)
+        logger.info(f"Got {len(my_trades)} trade(s).")
+        trades.extend(my_trades)
+        if len(my_trades) == 0:
+            break
+        ts = my_trades[0]['timestampms']
+        payload_nonce += 1000
+        time.sleep(0.5)
+    # Dedup
+    trades = sorted(trades, key = lambda x: x['timestampms'])
+    unique_trades = {tr['tid'] : tr for tr in trades}
+    unique_trades = list(tr for _, tr in unique_trades.items())
+
+    # Now turn into a CSV
+    data_file = open(os.path.join(data_dir, "trades.csv"), 'w')
+    csv_writer = csv.writer(data_file)
+    header = None
+    for tr in unique_trades:
+        if not header:
+            header = tr.keys()
+            csv_writer.writerow(header)
+        csv_writer.writerow(tr.values())
+    data_file.close()
+
+
+def get_balances_and_prices(api_key, api_secret, data_dir):
+    payload_nonce =  int(time.time()*1000)
+
+    payload =  {"request": "/v1/balances", "nonce": str(payload_nonce)}
+    encoded_payload = json.dumps(payload).encode()
+    b64 = base64.b64encode(encoded_payload)
+    signature = hmac.new(api_secret, b64, hashlib.sha384).hexdigest()
+
+    request_headers = get_request_headers(api_key, b64, signature)
+    response = requests.post(BALANCES_URL, headers=request_headers)
+    response.raise_for_status()
+
+    balances = response.json()
+    if 'result' in balances:
+        raise Exception(balances)
+    logger.info(f"Got balances. Found {len(balances)} currencies.")
+
+    # Prices
+    tickers = [b['currency'] + "USD" for b in balances if b['currency'] != 'USD']
+    prices = {}
+    for t in tickers:
+        response = requests.get(TICKERS_URL+ "/"+ t.lower())
+        response.raise_for_status()
+        obj = response.json()
+        price = (float(obj['ask']) + float(obj['bid']))/2
+        currency = obj['symbol'][:-3] # BTCUSD -> BTC
+        prices[currency] = price
+        time.sleep(0.2)
+    logger.info("Got prices")
+
+    # Merge and Date
+    for bal in balances:
+        bal['price'] = bal['price'] = prices[bal['currency']]\
```

Comment:
No reason to duplicate `bal['price'] =` twice here

Reply:
yup

Reply:
This duplication is still present